### DISCUSSION TITLE
Allowing a guard or fairing to set data available in templates?

### DISCUSSION THREAD

**Discussion by inabahare** [rwf2/Rocket#2147] [Questions]
Or maybe Coming from a mainly express.js background, I can't seem to find out if this is possible. 

Imagine you have a page where after a user as logged in, you need to show their name in the top right corner. Where with handlebars you could do something like:

```hbs
{{#if user}}
<p>{{user.username}}
{{/if}}
```

which you would have in your template. But how would you achieve this? In express I'm used to [res.locals](http://expressjs.com/en/api.html#res.locals) which is a variable you can assign data to in a middleware and have available in your template.

But I can't really figure out what I would do here? I already have my guard for authentication, can it set something?

**Comment by the10thWiz** [comment]
The short answer is no, I don't believe this is possible. That being said, I believe you could include the authentication guard (or some related struct) in your template context. The downside is you will have to explicitly do this in every handler, but I don't think there is a simpler option.

```rust
#[get("/")]
fn index_page(user: User<'_>) -> Template {
  Template::render(IndexCtx {
    user, // could also be `user: user.render_ctx(),`
    ..
  })
}
```

**Comment by inabahare** [comment]
Yeah I had the idea to do something like that, or have an authenticated version of the public routes that just pass the user object to the template.But nothing really seems like it would be ideal tbh

**Reply by the10thWiz** [reply]
I'm going to look into the ability for a fairing to add template context, but I'm not sure how it should work.

The major issue is the template context itself. How should Rocket attach the extra information? The obvious answer would be to place the extra context in the request's local_cache, but this doesn't handle allowing multiple extra context objects, and it also doesn't handle conflicts, i.e. two fairings add objects with the same name (but different information), or the user code defining the same info. This also breaks the `Template::show` method, since it doesn't have access to the request's local cache.

I think this could be done, but it's not a trivial operation. I think requiring the user to specify that they want to add the user struct is fine - It's a little bit more code, but it makes it easier to tell what is available in the template context.
